File: Python/438.找到字符串中所有字母异位词.py
from typing import List

class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        ans: List[int] = []
        if len(s) < len(p):
            return ans
        sts: List[int] = [0] * 26
        stp: List[int] = [0] * 26
        for i in range(len(p)):
            sts[ord(s[i]) - ord('a')] += 1
            stp[ord(p[i]) - ord('a')] += 1
        if sts == stp:
            ans.append(0)
        left: int = 0
        while left + len(p) < len(s):
            sts[ord(s[left]) - ord('a')] -= 1
            sts[ord(s[left + len(p)]) - ord('a')] += 1
            left += 1
            if (sts == stp):
                ans.append(left)
        return ans

File: 438.找到字符串中所有字母异位词.cc
#include <vector>
#include <string>
using std::vector;
using std::string;

class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        vector<int> ans;
        if (s.size() < p.size())
            return ans;
        vector<int> stp(26);
        vector<int> sts(26);
        for (int i = 0; i < p.size(); ++i) {
            ++stp[p[i] - 'a'];
            ++sts[s[i] - 'a'];
        }
        if (stp == sts) {
            ans.push_back(0);
        }
        int left = 0;
        while (left + p.size() < s.size()) {
            --sts[s[left] - 'a'];
            ++sts[s[left + p.size()] - 'a'];
            ++left;
            if (sts == stp) {
                ans.push_back(left);
            }
        }
        return ans;
    }
};
